docs/resource/README.md

```diff
@@ -102,7 +102,105 @@ as specified in the [Resource SDK specification](https://github.com/open-telemet
+(e.g. instances of a horizontally scaled service). It is preferable for the ID to be persistent
+and stay the same for the lifetime of the service instance, however it is acceptable that
+the ID is ephemeral and changes during important lifetime events for the service
+(e.g. service instance restarts).
+
+If the instance has no inherent unique ID that can be used as the value of this attribute,
+implementations MAY to generate a random Version 1 or Version 4
+[RFC 4122](https://www.ietf.org/rfc/rfc4122.txt) UUID. When enough data is available,
+implementations SHOULD use Version 5 and MUST use the following UUID as the namespace:
+`4d63009a-8d0f-11ee-aad7-4c796ed8e320`.
+
+UUIDs are typically recommended, as we only need an opaque yet reproducible value for
+the purposes of identifying a service instance. Similar to what can be seen in the man page for the
+[`/etc/machine-id`](https://www.freedesktop.org/software/systemd/man/machine-id.html) file,
+the underlying data, such as pod name and namespace should be treated as
+confidential by this algorithm, being the user's choice to expose it or not via another resource attribute.
+
+When a UUID v5 is generated, the input MUST be prefixed with
+`${telemetry.sdk.name}.${telemetry.sdk.language}.${service.namespace}.${service.name}`, followed by the
+the workload identifier, which should tentatively stable. This means that the same service yields the
+same UUID if the same identifier (`host.id`, `/etc/machine-id`, and so on) remains the same. It would still
+yield different results for different services on the same host or namespace. When no namespaces or equivalent fields
+are available, the prefix MUST then be `${telemetry.sdk.name}.${telemetry.sdk.language}.${service.name}`.
+
+Users running their services on platforms such as Kubernetes are encouraged to explicitly set the
+`service.instance.id` using their existing automation, or by setting a value that can be used for a
+consistent value, such as `container.id`. Similarly, users of application servers such as `unicorn`
+are encouraged to set the `service.instance.id` on a per-worker basis.
+
+SDKs MUST use the following algorithm when generating `service.instance.id`:
+
+- If the user has provided a `service.instance.id`, via environment
+  variable, configuration or custom resource detection, this will
+  always be used and honored over generated IDs.
+- When any of the below combinations of resource attribute are provided, they MUST be used as the input
+  for generating a UUID v5 following the prefix mentioned above. The values within each combination MUST be separated with dots:
+  * `container.id`, resulting in the input `${telemetry.sdk.name}.${telemetry.sdk.language}.${service.namespace}.${service.name}.${container.id}`
+  * `k8s.namespace.name`/`k8s.pod.name`/`k8s.container.name`, resulting in the input
+  `${telemetry.sdk.name}.${telemetry.sdk.language}.${k8s.namespace.name}.${service.name}.${k8s.pod.name}.${k8s.container.name}`
```

Comment:
Would these `k8s.*` attributes have to be user provided? I know the collector has a processor which can provide them, but I know at least in java there is no k8s resource provider. See list of available resource providers [here](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/resources/library/src/main/java/io/opentelemetry/instrumentation/resources).

If its not practical for the SDKs to have built in logic for detecting these attributes, does it make sense to include `service.instance.id` algorithm which depends on them?

Reply:
> I know the collector has a processor which can provide them

I'm not sure I'd rely on the collector to generate the service.instance.id. I'd prefer to have it closer to the workload, either by an operator or by the SDK.

> If its not practical for the SDKs to have built in logic for detecting these attributes

Almost all of them can be easily inferred by the SDKs:

- namespace name can be read from `/var/run/secrets/kubernetes.io/serviceaccount/namespace`
- pod name is typically the hostname. 

Although the best is to use the token file and base64 decode the payload:

- read /var/run/secrets/kubernetes.io/serviceaccount/token , which looks like this: `eyJhbGciOiJSUzI1NiIsImtpZCI6IjJkRW5VaFh3bGtBU05xSlZyS3ZSNXlUN3htQUFaTng5M1ZWWEdXcWRkdmcifQ.eyJhdWQiOlsiaHR0cHM6Ly9rdWJlcm5ldGVzLmRlZmF1bHQuc3ZjLmNsdXN0ZXIubG9jYWwiLCJrM3MiXSwiZXhwIjoxNzMzNTY1ODMzLCJpYXQiOjE3MDIwMjk4MzMsImlzcyI6Imh0dHBzOi8va3ViZXJuZXRlcy5kZWZhdWx0LnN2Yy5jbHVzdGVyLmxvY2FsIiwia3ViZXJuZXRlcy5pbyI6eyJuYW1lc3BhY2UiOiJkZWZhdWx0IiwicG9kIjp7Im5hbWUiOiJzaGVsbC1kZW1vIiwidWlkIjoiNDNiNDVjNDgtNzQwNC00NTAxLWI4MWItMDk3OWMyNjE0ZGYzIn0sInNlcnZpY2VhY2NvdW50Ijp7Im5hbWUiOiJkZWZhdWx0IiwidWlkIjoiZjJlNDgwYzgtYzAzOS00MjY4LTlhYTctMzUxN2VlOWE0NDEzIn0sIndhcm5hZnRlciI6MTcwMjAzMzQ0MH0sIm5iZiI6MTcwMjAyOTgzMywic3ViIjoic3lzdGVtOnNlcnZpY2VhY2NvdW50OmRlZmF1bHQ6ZGVmYXVsdCJ9.Cn6WCtdoNc9BOUtCSHrEh729pQK1iG2L5dciNUJnVzvgDN12SPWVB0BmZ9YmzwCSzKqgmi0kyRAyCGaYs-2uyLl9h1mjwiEqAoN89d0vLExheAxOz1x76jAJWhvMxr-nzF-_qbBEOnypHzNCoBVMYNXhwRR36ipoIhst4LwDoLAqEKKU595jQinlid7X3GVOWx41HF3V3JLHmXLLW-RqILouf_GPdEBqG6DFB-sWMBLa3PE-Rt5d37786PgsC6UGORRjCLfN04bgbYL1YGKAULvgZzFEviyRC3s5KNZG1xhQBv4P0lhJJWOwKPvazXCUt5bAytTakaHBGzH_Pld7mw`
- split this token into three parts at each dot, get the part at [1] and base64 decode it. In the example above, it would yield this: `{"aud":["https://kubernetes.default.svc.cluster.local","k3s"],"exp":1733565833,"iat":1702029833,"iss":"https://kubernetes.default.svc.cluster.local","kubernetes.io":{"namespace":"default","pod":{"name":"shell-demo","uid":"43b45c48-7404-4501-b81b-0979c2614df3"},"serviceaccount":{"name":"default","uid":"f2e480c8-c039-4268-9aa7-3517ee9a4413"},"warnafter":1702033440},"nbf":1702029833,"sub":"system:serviceaccount:default:default"}` 

The container name cannot be inferred by the SDKs though. I can be provided by the Operator, which can also calculate the service.instance.id directly as well. People not using the operator might still want to have the SDKs generate the service.instance.id, but for that, they'd need to specify the container.name. And if they are specifying the container name, perhaps they should set the other ones explicitly as well.

In summary: the SDKs can infer most of the data, but container.name is still missing. I'd be fine in having the SDKs not implement this part of the algorithm, but I'd prefer if they could still generate a proper service.instance.id if the container.name is provided explicitly.

Reply:
I think its probably fine to leave it as part of the algorithm. The code generating `service.instance.id` can look for this combination of resource attributes, and use them to generate if available. I suspect most of the time they won't be available, but simple enough if they are.